### PR TITLE
[CAMEL-18472] lock 'model' in 'removeRoute' first to avoid deadlock

### PR DESCRIPTION
**Pull request by gerdailger**
Obtain the lock on `model` before locking `camelContext` in `removeRoute` to avoid a potential deadlock with concurrent `addRoutes` calls.

### CODE REVIEW

File: core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        Supplier<Callable<Boolean>> addRouteSupplier = () -> {
+            Callable<Boolean> addRouteTask = () -> {
+                String routeId = UUID.randomUUID().toString();
+                context.addRoutes(new RouteBuilder() {
+                    @Override
+                    public void configure() throws Exception {
+                        String endpointUri = "tmp:start_" + routeId;
+                        MySlowEndpoint mySlowEndpoint = new MySlowEndpoint(endpointUri, context);
+                        from(mySlowEndpoint).id(routeId).to("mock:result");
+                        log.info("Route " + endpointUri + " configured");
```
Comment (orpiske) [inline core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java:65]: Can you please use log markers here? As in:

log.info("Route {} configured", endpointUri);

**Review by orpiske — COMMENTED** [review]
Thanks for the contribution. I've left one advisory comment. 

@davsclaus @oscerd I'm not sure if you are planning to test this one, but FYI I'll trigger some manual tests on this as the core code is not tested on PRs.

File: core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java (hunk 1 above)
Comment (gerdailger) [reply core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java]: Oh yes of course, thank you!

**Review by orpiske — APPROVED** [review]
Manually tested with: 

- camel-ftp
- camel-jms
- camel-kafka
- camel-couchbase
- core/camel-core 

There were no issues.

File: core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java (hunk 1 above)
Comment (orpiske) [reply core/camel-core/src/test/java/org/apache/camel/issues/DynamicallyConcurrentlyAddRoutesTest.java]: Thanks!

**Review by oscerd — APPROVED** [review]

**Comment by oscerd** [comment]
Let's wait for @davsclaus

**Comment by davsclaus** [comment]
The method `disposeModel` is used when in lightweight mode to clear the definition classes for runtimes where you do not dynamic add/remove routes.

So even if model is null then an exception would be thrown so that should be okay

**Comment by davsclaus** [comment]
LGTM

**Review by davsclaus — APPROVED** [review]

**Pull request merged**